beam/core/src/main/java/cz/o2/proxima/beam/io/ProximaIO.java

```diff
@@ -0,0 +1,112 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.beam.io;
+
+import com.google.common.annotations.VisibleForTesting;
+import cz.o2.proxima.annotations.Experimental;
+import cz.o2.proxima.direct.core.DirectDataOperator;
+import cz.o2.proxima.repository.RepositoryFactory;
+import cz.o2.proxima.storage.StreamElement;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+
+/** IO connector for Proxima platform. */
+@Experimental
+@Slf4j
+public class ProximaIO {
+
+  private ProximaIO() {
+    // No-op.
+  }
+
+  /**
+   * Write {@link StreamElement stream elements} into proxima using {@link DirectDataOperator}.
+   *
+   * @param repositoryFactory Serializable factory for Proxima repository.
+   * @return Write transform.
+   */
+  public static Write write(RepositoryFactory repositoryFactory) {
+    return new Write(repositoryFactory);
+  }
+
+  /**
+   * Transformation that writes {@link StreamElement stream elements} into proxima using {@link
+   * DirectDataOperator}.
+   */
+  public static class Write extends PTransform<PCollection<StreamElement>, PDone> {
+
+    private final RepositoryFactory repositoryFactory;
+
+    private Write(RepositoryFactory repositoryFactory) {
+      this.repositoryFactory = repositoryFactory;
+    }
+
+    @Override
+    public PDone expand(PCollection<StreamElement> input) {
+      input.apply("Write", ParDo.of(new WriteFn(repositoryFactory)));
+      return PDone.in(input.getPipeline());
+    }
+  }
+
+  static class WriteFn extends DoFn<StreamElement, Void> {
+
+    private final RepositoryFactory repositoryFactory;
+
+    private transient DirectDataOperator direct;
+
+    WriteFn(RepositoryFactory repositoryFactory) {
+      this.repositoryFactory = repositoryFactory;
+    }
+
+    @VisibleForTesting
+    DirectDataOperator getDirect() {
+      return direct;
+    }
+
+    @Setup
+    public void setUp() {
+      direct = repositoryFactory.apply().getOrCreateOperator(DirectDataOperator.class);
+    }
+
+    @ProcessElement
+    public void processElement(@Element StreamElement element) {
+      direct
+          .getWriter(element.getAttributeDescriptor())
+          .orElseThrow(
+              () ->
+                  new IllegalArgumentException(
+                      String.format("Missing writer for [%s].", element.getAttributeDescriptor())))
+          .write(
```

Comment:
This is definitely not exactly once. We may do the following:
 a) BufferUntilCheckpoint, or
 b) synchronize the write, or
 c) mark writers as idempotent (e.g. cassandra) vs. non-idempotent (kafka)
We are probably free to merge this as it is, but we should file an issue to fix that appropriately. I prefer option c) with user decision on how to solve the problem for non-idempotent sinks (either a) or b) or ignore that knowing that the write might - and will - cause duplicates).

Reply:
I filled https://github.com/O2-Czech-Republic/proxima-platform/issues/330. We can merge this.